pkg/ingester/ingester.go

```diff
@@ -2299,6 +2300,81 @@ func (i *Ingester) RingHandler() http.Handler {
 	return i.lifecycler
 }
 
+func (i *Ingester) AddBackfillFile(stream client.Ingester_AddBackfillFileServer) error {
+	if err := i.checkRunning(); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+
+	level.Info(i.logger).Log("msg", "processing request to add backfill file")
+
+	f, err := os.CreateTemp("", "")
```

Comment:
Why going through a local temporary file? Can't we directly feed the `io.Reader` passed to `bkt.Upload()`?

Reply:
In this method we don't have an `io.Reader`? Here we instead receive a stream, from which we read individual messages each containing a chunk of bytes.

Reply:
Come to think of it, maybe I could wrap the gRPC stream in an `io.Reader` implementation?

Reply:
We don't have `io.Reader`, because we're going through gRPC. Original idea was to implement backfilling as HTTP API, where we would have `io.Reader`.

Reply:
@pracucci @pstibrany I've rewritten the bucket uploading logic to use a stream wrapper implementing `io.Reader` instead of a temporary file. From preliminary testing, it works (the expected number of bytes gets written).